sparse_strips/vello_hybrid/src/render.rs

```diff
@@ -0,0 +1,268 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Basic render operations.
+
+use crate::gpu::{GpuStrip, RenderData};
+use crate::{RenderTarget, Renderer};
+use kurbo::{Affine, BezPath, Cap, Join, Rect, Shape, Stroke};
+use peniko::color::palette::css::BLACK;
+use peniko::{BlendMode, Compose, Fill, Mix};
+use vello_common::coarse::{WIDE_TILE_WIDTH, Wide};
+use vello_common::flatten::Line;
+use vello_common::paint::Paint;
+use vello_common::pixmap::Pixmap;
+use vello_common::strip::{STRIP_HEIGHT, Strip};
+use vello_common::tile::Tiles;
+use vello_common::{flatten, strip};
+
+/// Default tolerance for curve flattening
+pub(crate) const DEFAULT_TOLERANCE: f64 = 0.1;
+
+/// A render context for hybrid CPU/GPU rendering.
+///
+/// This context maintains the state for path rendering and manages the rendering
+/// pipeline from paths to strips that can be rendered by the GPU.
+#[derive(Debug)]
+pub struct RenderContext {
+    pub(crate) width: usize,
+    pub(crate) height: usize,
+    pub(crate) wide: Wide,
+    pub(crate) alphas: Vec<u32>,
+    pub(crate) line_buf: Vec<Line>,
+    pub(crate) tiles: Tiles,
+    pub(crate) strip_buf: Vec<Strip>,
+    pub(crate) paint: Paint,
+    pub(crate) stroke: Stroke,
+    pub(crate) transform: Affine,
+    pub(crate) fill_rule: Fill,
+    pub(crate) blend_mode: BlendMode,
+}
+
+impl RenderContext {
+    /// Create a new render context with the given width and height in pixels.
+    pub fn new(width: u16, height: u16) -> Self {
+        // TODO: Use u16 for width/height everywhere else, too.
+        let wide = Wide::new(width.into(), height.into());
+
+        let alphas = vec![];
+        let line_buf = vec![];
+        let tiles = Tiles::new();
+        let strip_buf = vec![];
+
+        let transform = Affine::IDENTITY;
+        let fill_rule = Fill::NonZero;
+        let paint = BLACK.into();
+        let stroke = Stroke {
+            width: 1.0,
+            join: Join::Bevel,
+            start_cap: Cap::Butt,
+            end_cap: Cap::Butt,
+            ..Default::default()
+        };
+        let blend_mode = BlendMode::new(Mix::Normal, Compose::SrcOver);
+
+        Self {
+            width: width.into(),
+            height: height.into(),
+            wide,
+            alphas,
+            line_buf,
+            tiles,
+            strip_buf,
+            transform,
+            paint,
+            fill_rule,
+            stroke,
+            blend_mode,
+        }
+    }
+
+    /// Fill a path with the current paint and fill rule.
+    pub fn fill_path(&mut self, path: &BezPath) {
+        flatten::fill(path, self.transform, &mut self.line_buf);
+        self.render_path(self.fill_rule, self.paint.clone());
+    }
+
+    /// Stroke a path with the current paint and stroke settings.
+    pub fn stroke_path(&mut self, path: &BezPath) {
+        flatten::stroke(path, &self.stroke, self.transform, &mut self.line_buf);
+        self.render_path(Fill::NonZero, self.paint.clone());
```

Comment:
Hmm, so we finalise the render for each path as we go? Interesting. My gut reaction is that it is simpler, but has costs around how much batching we are doing, as well as blocking the UI thread on things which should be on the render thread (slightly moot at the moment as we do that all on a single thread)